src/client/datascience/notebook/cellOutputs.ts

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { nbformat } from '@jupyterlab/coreutils';
+import cloneDeep from 'lodash/cloneDeep';
+import { NotebookCellOutput, NotebookCellOutputItem } from 'vscode';
+import { concatMultilineString, splitMultilineString } from '../../../datascience-ui/common';
+import { traceWarning } from '../../common/logger';
+import { sendTelemetryEvent } from '../../telemetry';
+import { Telemetry } from '../constants';
+import { CellOutputMimeTypes } from './types';
+
+/**
+ * Metadata we store in VS Code cell output items.
+ * This contains the original metadata from the Jupyter Outputs (nbformat.IOutput).
+ */
+export interface CellOutputMetadata {
+    /**
+     * Cell output metadata (found in nbformat.IDisplayData.metadata, etc)
+     */
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    metadata?: any;
+    /**
+     * Transient data from Jupyter.
+     * Cell output metadata (found in nbformat.IDisplayData.transient, etc)
+     */
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    transient?: {
+        /**
+         * This is used for updating the output in other cells.
+         * We don't know of others properties, but this is definitely used.
+         */
+        display_id?: string;
+    } & any;
+    /**
+     * Cell output metadata (found in nbformat.IExecuteResult.execution_count, etc)
+     */
+    executionCount?: nbformat.IExecuteResult['ExecutionCount'];
+    streamName?: nbformat.StreamType;
+}
+
+abstract class BaseNotebookCellOutput<T extends nbformat.OutputType> extends NotebookCellOutput {
+    public get cellOutputMetadata() {
+        return this.metadata as CellOutputMetadata | undefined;
+    }
+    constructor(
+        outputs: NotebookCellOutputItem[],
+        public readonly outputType: T,
+        cellOutputMetadata?: CellOutputMetadata
+    ) {
+        super(outputs, cellOutputMetadata);
```

Comment:
This is new VSC API, they now allow us to store metadta against the output, previous we could only do this in the output item (and we had to resort to using private map to support `transient.display_id`)..


Now Document, cells, outupts, output items - everything in notebooks support metadata...
We are using interfaces to ensur ethey are strongly typed.